### PR TITLE
Implement Canvas pixel manipulation

### PR DESCRIPTION
**Pull request by ebalint**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/4639)

### CODE REVIEW

**Comment by jdm** [comment]
Looking good! I left some comments that are mostly about making the code more readable :)

File: components/canvas/canvas_paint_task.rs [hunk 1]
```diff
@@ -85,4 +94,101 @@ impl CanvasPaintTask {
...
+                    dirty_rect.size.height = -dirty_rect.size.height;
+                }
+                new_image_data_rect.intersection(&dirty_rect)
+            },
+            None => Some(new_image_data_rect)
+        };
+
+        if let Some(_) = new_dirty_rect {
+            let moved_dirty_rect = Rect(new_dirty_rect.unwrap().origin.add(image_data_rect.origin),
+                                        new_dirty_rect.unwrap().size).intersection(&Rect(Point2D(0i32, 0i32),
```
Comment (jdm) [inline components/canvas/canvas_paint_task.rs]: If we use `if let Some(new_dirty_rect) = new_dirty_rect {` above, we can get rid of these unwrap calls.

**Comment by jdm** [comment]
Oh, this is totally ready to merge if you squash the commits now :)

**Comment by ebalint** [comment]
I squashed :)

**Comment by jdm** [comment]
```
thread '/Users/servo/buildbot/slave/mac2/build/tests/content/test_interfaces.html' panicked at 'ALERT: TEST-UNEXPECTED-FAIL | ImageData is in [object Object]: If this is failing: DANGER, are you sure you want to expose the new interface ImageData to all webpages as a property on the window? Do not make a change to this file without review from jdm or Ms2ger for that specific change!', /Users/servo/buildbot/slave/mac2/build/tests/contenttest.rs:122
```
```
/html/semantics/embedded-content/the-canvas-element/initial.colour.html
-----------------------------------------------------------------------
PASS expected FAIL Initial state is transparent black
/workers/semantics/interface-objects/001.worker
-----------------------------------------------
PASS expected FAIL The ImageData interface object should be exposed.
```

**Comment by jdm** [comment]
I'm rebasing this and annotating the passing tests.

**Pull request closed without merging**